functions/go/set-labels/transformer/labels.go

```diff
@@ -0,0 +1,168 @@
+		p.NewLabels = o.NestedStringMapOrDie("labels")
+		if len(p.NewLabels) == 0 {
+			return fmt.Errorf("failed to configure function: input label list cannot be empty")
+		}
+	default:
+		return fmt.Errorf("unknown functionConfig Kind=%v ApiVersion=%v, expect `%v` or `ConfigMap`",
+			o.GetKind(), o.GetAPIVersion(), fnConfigKind)
+	}
+	// add default fields
+	if err := p.addDefaultLabelFields(); err != nil {
+		return err
+	}
+	// add additional fields
+	if o.IsGVK(fnConfigGroup, fnConfigAPIVersion, fnConfigKind) {
+		arr, exist, err := o.NestedSlice("additionalLabelFields")
+		if err != nil {
+			return fmt.Errorf("`additionalLabelFields` format is wrong")
+		}
+		if exist {
+			var addFields []FieldSpec
+			for _, sub := range arr {
+				var curField FieldSpec
+				if err := sub.As(&curField); err != nil {
+					return err
+				}
+				addFields = append(addFields, curField)
+			}
+			p.AdditionalLabelFields = append(p.AdditionalLabelFields, addFields...)
+		}
+	}
+	return nil
+}
+
+func (p *LabelTransformer) Transform(objects fn.KubeObjects) error {
+	for _, o := range objects {
+		for _, sp := range p.AdditionalLabelFields {
+			if (sp.Group == "" && sp.Version == "" && sp.Kind == "") || o.IsGVK(sp.Group, sp.Version, sp.Kind) {
+				newResult := fn.Result{
+					Message:     "Replace labels",
+					Severity:    "INFO",
+					ResourceRef: nil, // TODO: initialize in kyaml
+					Field: &fn.Field{
+						Path:          sp.Path,
+						CurrentValue:  nil, // values to be updated in setLabel()
+						ProposedValue: nil, // values to be updated in setLabel()
+					},
+					File: &fn.File{
+						Path:  o.PathAnnotation(),
+						Index: o.IndexAnnotation(),
+					},
+					Tags: nil,
+				}
+				err := updateLabels(o, sp.Path, p.NewLabels, sp.CreateIfNotPresent, &newResult)
+				if err != nil {
+					return err
+				}
+				p.Results = append(p.Results, &newResult)
+			}
+		}
+	}
+	return nil
+}
+
+// TODO: Is there any helper function insdie fn package to help with copy map
+func copyStringMap(m map[string]string) map[string]string {
+	if m == nil {
+		return nil
+	}
+	newMap := make(map[string]string)
+	for k, v := range m {
+		newMap[k] = v
+	}
+	return newMap
+}
+
+func updateLabels(o *fn.KubeObject, fieldPath string, newLabels map[string]string, create bool, result *fn.Result) error {
+	paths := strings.Split(fieldPath, "/")
+
+	oldLabels, exist, err := o.NestedStringMap(paths...)
+	if err != nil {
+		return err
+	}
+	if !exist {
+		oldLabels = make(map[string]string)
+	}
+	result.Field.CurrentValue = copyStringMap(oldLabels)
+	replaceLabels(oldLabels, newLabels, create)
+	result.Field.ProposedValue = copyStringMap(oldLabels)
+	o.SetNestedStringMapOrDie(oldLabels, paths...)
+	return nil
+}
+
+// replaceLabels replace old labels map with new labels map according to create
+// oldLabels must not be nil
+func replaceLabels(oldLabels map[string]string, newLabels map[string]string, create bool) {
```

Comment:
This replacement is not going to work, since this is not modifying the original object.
You should use something like `o.Set()` to modify the original object.

Reply:
I have now changed my code to use `o.SetNestedString()`. It is more concise now. Thank you for your advice!